### PR TITLE
Cleanup the use of ExternalID as it is deprecated

### LINKED ISSUE #60466: Confusion around node name: providerID, externalID
Currently both node_controller in cloudprovider and node_life_cycle controller checks whether node exists in cloudprovider or not. They are using different ways of checking the cloudprovider.
node_controlelr: InstanceExistsByProviderID by passing node's providerID. If there is an error, it will try to use ExternalID()
node_lifecycle_controller: Uses ExternalID() by passing nodeName.

However, according to this PR https://github.com/kubernetes/kubernetes/pull/7775, ExternalID is marked as deprecated three years ago. Should we continue to use this way?

cc @yujuhong @Random-Liu

### PR DESCRIPTION
**Pull request by adnavare**
The patch removes ExternalID usage from node_controller
and node_lifecycle_oontroller. The code instead uses InstanceID
which returns the cloud provider ID as well.

fixes #60466

### CODE REVIEW

**Comment by adnavare** [comment]
Hi Folks, I have changed the code to remove ExternalID from node_controller and node_lifecycle_controller. But as ExternalID is deprecated and if not needed, I can go ahead  to remove the implementation of ExternalID() from each provider and also delete the field from instances type defined in cloud.go. I wanted to know the feedback from you guys, thanks.

**Comment by adnavare** [comment]
Can someone please approve the PR to be tested? Thanks

**Comment by deads2k** [comment]
/assign @jhorwit2 
/assign @aveshagarwal 

I think these reviewers are closer than me.

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/60692)

**Review by jhorwit2 — CHANGES_REQUESTED** [review]
Thanks for the PR @adnavare! Sadly, I don't think we can move from `ExternalID` to `InstanceID` this easily since the interface contract is different. `ExternalID` explicitly says 

> // Note that if the instance does not exist or is no longer running, we must return ("", cloudprovider.InstanceNotFound)

and the `InstanceID` method does not. From what I can tell quickly looking through some cloud provider implementations this change would cause nodes to not be deleted since some clouds only return the error mentioned above in `ExternalID`.

I'd suggest writing a brief proposal and sending it to the working groups [email group](https://github.com/kubernetes/community/tree/master/wg-cloud-provider#contact) to get opinions/feedback from as many interested parties as possible.

**Comment by jhorwit2** [comment]
An interesting point about the deprecation... it seems that the `ExternalID`method on the `Instances` interface does **not** mention it's deprecated; however, the implementations of specific clouds do mention it is deprecated in their method doc.

**Comment by adnavare** [comment]
Hi @jhorwit2 : Thanks for the review. Yes, true but so far I found only two places where the error returned is used to check future condition and that is in node_controller.go, controller_utils.go and in two unit tests - photon_test.go and vsphere_test.go. 
In those cases what can be done is check if the ID returned is nil or not, as InstanceID() returns ID or nil. 
But yes, I did not change in the code right now because was waiting for feedback from folks.

**Comment by adnavare** [comment]
@jhorwit2: Yes indeed, i noticed that, deprecation note about ExternalID is not present on "Instances" interface

**Comment by adnavare** [comment]
Yes, due to InstanceID not returning bool,error instead returning bool,nil, I had to change a unit test. I will update based upon the feedback from folks later. I will send out an email to mailing list today asking for feedback whether to replace ExternalID entirely. Thank you.

**Comment by adnavare** [comment]
@jhorwit2 @wlan0: Can you get your eye over the cod, thanks!

**Comment by jhorwit2** [comment]
@adnavare there is more code that uses the ExternalID method like the [kubelet](https://github.com/kubernetes/kubernetes/blob/8182501bb5fd4602136a95bf23d70cffb72e1600/pkg/kubelet/kubelet_node_status.go#L305). If we're OK with completely removing the calls then we should also remove it from the Instances cloud provider interface.

**Comment by adnavare** [comment]
@jhorwit2: Yes i agree. As from this week's discussion it is decided in the next week's meeting, representative for each cloud provider will give an idea whether they want to remove ExternalID completely. If so, then I will go ahead and remove the ExternalID method from Instances cloud provider interface.

**Comment by wlan0** [comment]
> As from this week's discussion it is decided in the next week's meeting, representative for each cloud provider will give an idea whether they want to remove ExternalID completely. 

@adnavare We don't have a channel for obtaining this information yet. Please go ahead and make the change. We can tag the cloud provider reps on the PR once the change is complete and ready to be reviewed by them.

**Comment by adnavare** [comment]
@wlan0: Okay. Sure

**Comment by adnavare** [comment]
I have updated few files to remove the usage of ExternalID. Also while making these changes, what I noticed is, ExternalID is also defined in NodeSpec inside apis/core/types.go and we refer that value in kubelet/kubelet_node_status.go. Will it make sense to add InstanceID instead of ExternalID in NodeSpec structure?

**Comment by adnavare** [comment]
the checks are failing because of the ExternalID reference in kubelet/node_status.go

File: pkg/cloudprovider/cloud.go
Comment (andrewsykim) [inline pkg/cloudprovider/cloud.go]:
```go
// Note that if the instance does not exist or is no longer running, we must return ("", cloudprovider.InstanceNotFound)
```
this is a pretty important comment we should keep. Can we add it to the comments for InstanceID?

Comment (andrewsykim) [reply pkg/cloudprovider/cloud.go]: We also can't remove the call to ExternalID without knowing that all clouds are able to return `cloudprovider.InstanceNotFound)` from InstanceID()

Comment (adnavare) [reply pkg/cloudprovider/cloud.go]: Hi @andrewsykim, thanks for the review. 
1. Yes I will add the comment for the InstanceID. 
2. Yes I have confirmed that all the cloud providers using that I have changed in this PR to use InstanceID indeed returns InstanceNotFound error. Some of the cloud providers (photon, ovirt) also test the error in their unit tests. So, that is taken care of.

**Comment by adnavare** [comment]
Also I noticed I haven't removed google cloud provider's ExternalID(). But I checked the code, and gce_instances.go have same code inside ExternalID() and InstanceID(), so won't be any complication in removing the method.

**Comment by andrewsykim** [comment]
@adnavare after reviewing the PR again, I have concerns around how AWS uses `ExternalID()` and `InstanceID()`. It seems `ExternalID()` returns the instances id and `InstanceID()` returns `/az/instance-id` (seems counter-intuitive ). Changing the `ExternalID` call to `InstanceID` would cause `node.Spec.ExternalID` to change between versions which might cause issues depending on what is consuming it. Would be good to get a review from someone in @kubernetes/sig-aws-misc on this.

**Comment by adnavare** [comment]
hmm, I see. So, return value of InstanceID() is consumed by ensureNodeExistsByProviderIDorExternalID(), existsInCLoudProvider() and node.Spec.ExternalID. In the first two, the values are ignored and we just return true/false. With regards, to the last one, do you think adding a logic to check presence of availability zone in the string? And if it exists, use mapToAwsInstance()?

**Comment by liggitt** [comment]
cc @mikedanese

File: pkg/cloudprovider/providers/aws/aws.go
Comment (mikedanese) [inline pkg/cloudprovider/providers/aws/aws.go]: cc @justinsb

**Comment by mikedanese** [comment]
ExternalID aliased to InstanceID
* azure
* cloudstack
* vsphere

ovirt InstanceID is ExternalID with a leading `/`

photon InstanceID and ExternalID are copies of the same function

gce InstanceID is `<project>/<zone>/<instance>` and ExternalID is `<instance>`

aws InstanceID is `/<zone>/<instance>` and ExternalID is `<instance>`

**Comment by mikedanese** [comment]
The kubelet change is safe, and we plan on removing that field entirely before 1.11. @tallclair for kubelet approval:

/assign @tallclair

**Comment by mikedanese** [comment]
release note none. we can roll the release note into #61877

**Comment by adnavare** [comment]
Please let me know if any AR is needed from my side on this PR to get it going. Thank you all for reviews and feedback.

File: pkg/cloudprovider/cloud.go
Comment (justinsb) [reply pkg/cloudprovider/cloud.go]: This doesn't appear to have been done.  I sent #64367 to do so

**Pull request merged**
